### PR TITLE
FlatGeobuf: implement spec v2

### PR DESCRIPTION
**Pull request by bjornharrtell**
This adapts to FlatGeobuf v2 which aims to align better with WKB geometry type enum, supports GeometryCollection, curves and some other types. While FlatGeobuf could be extended somewhat without breaking backward compatibility, this time I believe it needs to be breaking and I suppose/hope that is still be acceptable at this point.

Upstream work is done in https://github.com/bjornharrtell/flatgeobuf/pull/14.

@rouault if you have any opinions and/or feedback it would of course be appreciated.

### CODE REVIEW

**Comment by rouault** [comment]
> This is experimental work in progress to adapt to FlatGeobuf v2 which aims to align better with WKB geometry type enum, supports GeometryCollection, curves and some other types.

+1

> I suppose/hope that is still be acceptable at this point.

Yes, no problem for me to break backward compatibility while GDAL 3.1 isn't released (~ April 2020).

**Comment by bjornharrtell** [comment]
Some sanitizer problems remain but I'm not sure what is causing it. I find these lines in the build job log:

```
ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_2 flatbuffers/flatbuffers.h:91:25: runtime error: load of misaligned address 0x61a00016e0dc for type 'const double', which requires 8 byte alignment
0x61a00016e0dc: note: pointer points here
  04 00 00 00 00 00 00 20  ae 31 1d 41 00 00 00 20  40 2b 52 41 00 00 00 40  b5 65 1d 41 00 00 00 a0
              ^ 
ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_3 flatbuffers/base.h:386:23: runtime error: unsigned integer overflow: 18446744073709551615 + 1 cannot be represented in type 'unsigned long'
```

Any ideas @rouault?

File: autotest/ogr/ogr_flatgeobuf.py [hunk 1]
```diff
@@ -267,9 +265,21 @@ def test_ogr_flatgeobuf_3():
     wktRoundtrip('MULTIPOLYGON M (((35 10 1,45 45 2,15 40 3,10 20 4,35 10 5),(20 30 1,35 35 2,30 20 3,20 30 4)))')
     wktRoundtrip('MULTIPOLYGON ZM (((35 10 1 5,45 45 2 4,15 40 3 3,10 20 4 2,35 10 1 1),(20 30 4 1,35 35 3 2,30 20 2 3,20 30 1 4)))')
 
+    wktRoundtrip('GEOMETRYCOLLECTION (POINT (4 6),LINESTRING (4 6,7 10))')
```
Comment (rouault) [inline autotest/ogr/ogr_flatgeobuf.py]: extensive testing of all Z, M and ZM variants could be interesting. Actually I'm thinking that a function in autotest/pymod/ogrtest.py that would return a list of all those geometries could be useful to the testing of a number of geometries. Possibly with (include_z=True,include_m=True,include_zm=True) options depending on driver capabilities

**Comment by bjornharrtell** [comment]
I've tried to setup a local VM with Ubuntu 16.04 and replicate the way sanitizer does the build but as of yet I'm failing at that (build issues).

**Comment by bjornharrtell** [comment]
Was able to reproduce the sanitize setup locally and get the same lines but unfortunately I have no clue what is causing them (or even if they are the real problem why the build is failing).

**Comment by rouault** [comment]
As far as I can see, the issue is triggered by the `test_ogrsf -all_drivers` of test_ogr_basic_10. With just the log, I cannot determine if it is the unsigned integer overflow error that causes the test to fail (it doesn't for the ogr_flatgeobuf.py tests), or something else like a crash.
The  unsigned integer overflow error in the flatbuf code could probably be masked with a no_sanitize annotation if it is determined to be a safe overflow.

**Comment by bjornharrtell** [comment]
Ah, I see I did look in the wrong plac. I can reproduce the error in my VM and with CPL_DEBUG=ON I get the following output for `test_ogrsf -driver FlatGeobuf`:

```sh
INFO: TestCreate(FlatGeobuf).
GDAL: GDALDriver::Create(FlatGeobuf,/foo/test.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
GDAL: GDALClose(/foo/test.fgb, this=0x6120002e8c40)
GDAL: GDALDriver::Create(FlatGeobuf,/vsimem/test616.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
FlatGeobuf: geometryType: 0, hasZ: 0, hasM: 0, hasT: 0
FlatGeobuf: ResetReading
FlatGeobuf: Writing empty layer
FlatGeobuf: Wrote magicbytes (8 bytes)
FlatGeobuf: Created 5 columns for writing
FlatGeobuf: Wrote header (212 bytes)
GDAL: GDALClose(/vsimem/test616.fgb, this=0x6120002e78c0)
GDAL: GDALDriver::Create(FlatGeobuf,/vsimem/test617.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
FlatGeobuf: geometryType: 0, hasZ: 0, hasM: 0, hasT: 0
FlatGeobuf: Writing empty layer
FlatGeobuf: Wrote magicbytes (8 bytes)
FlatGeobuf: Created 0 columns for writing
FlatGeobuf: Wrote header (68 bytes)
GDAL: GDALClose(/vsimem/test617.fgb, this=0x6120002e8dc0)
GDAL: GDALDriver::Create(FlatGeobuf,/vsimem/test618.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
FlatGeobuf: geometryType: 0, hasZ: 0, hasM: 0, hasT: 0
FlatGeobuf: ResetReading
FlatGeobuf: Writing empty layer
FlatGeobuf: Wrote magicbytes (8 bytes)
FlatGeobuf: Created 5 columns for writing
FlatGeobuf: Wrote header (212 bytes)
GDAL: GDALClose(/vsimem/test618.fgb, this=0x6120002e7bc0)
GDAL: GDALDriver::Create(FlatGeobuf,/vsimem/test619.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
FlatGeobuf: geometryType: 0, hasZ: 0, hasM: 0, hasT: 0
FlatGeobuf: Writing empty layer
FlatGeobuf: Wrote magicbytes (8 bytes)
FlatGeobuf: Created 0 columns for writing
FlatGeobuf: Wrote header (68 bytes)
GDAL: GDALClose(/vsimem/test619.fgb, this=0x6120002e8f40)
GDAL: GDALDriver::Create(FlatGeobuf,/vsimem/test620.fgb,0,0,0,Unknown,(nil))
FlatGeobuf: Spatial index requested will write to temp file and do second pass on close
FlatGeobuf: geometryType: 1, hasZ: 0, hasM: 0, hasT: 0
flatbuffers/base.h:386:23: runtime error: unsigned integer overflow: 18446744073709551615 + 1 cannot be represented in type 'unsigned long'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior flatbuffers/base.h:386:23 in
```

One would like it to actually say something after the last line :) But I'll see what I can guess...

**Comment by rouault** [comment]
can you just merge master into your branch so as to get the Mac CI fix ?

**Comment by bjornharrtell** [comment]
Rebased and I no longer consider it experimental.

**Comment by rouault** [comment]
OK, I had a comment about maybe a more extensive testing of all geometry type variants : https://github.com/OSGeo/gdal/pull/2051#discussion_r351010423

**Comment by bjornharrtell** [comment]
@rouault ah yes, I kind of put it to the back of my head as a "nice to have" as it would make sense to make such tests across drivers but I agree it would be good to at least have it verified for this driver in this PR. I'll give it a shot in the coming days so it's fine to keep this open until further notice.

**Pull request merged**